Adds table valued functions support

neisbut opened the pull request:
Hello guys, this PR adds support for table valued stored functions. 
1. I've added a tiny TVF to context which can query Blog records with simple ILIKE.
2. Added all needed function declarations in BloggingContext.CreateModel
3. And fixed core a bit of course.

Please review and let me know what you think. Thanks!

dkabracadabra commented:
Where have you been 2 days ago:( I've done exactly the same yesterday. I could save a day wasted on investigation and debugging. Anyway THANK YOU!

neisbut commented:
Hi, @dkabracadabra. Sorry, I was collecting information :smile: Hope guyw will check this soon. We are keen to have this feature!

neisbut commented:
Hi @roji, is there a chance you will check this PR soon?

roji commented:
I'm really sorry but it's unlikely I'll be able to review this anytime soon... I know the EF6 provider hasn't received much attention recently. @Emill, if you have any time for this that would be great. Otherwise you guys will have to work with your own fork until I have more free time...

rwasef1830 commented:
Hello @neisbut, 
Could you rebase this PR on latest dev for review ?

neisbut commented:
Hi @rwasef1830, done!

rwasef1830 commented:
@neisbut When I try to run the test under a debugger, I hit many assert failures inside EntityFramework itself:

```
Debug.Assert(column.TypeUsage.EdmType.DataSpace == DataSpace.SSpace);
```

This happens in the constructor of ```ScalarPropertyMapping``` when ```dbModel.Compile()``` when running the unit test setup.

Although the test passes, the asserts mean that some core assumption inside EF is broken. I am not exactly familiar with the gory details of the CSpace and SSpace. Could you check it ?

Also, there is a project on GitHub https://github.com/Dixin/EntityFramework.Functions which seems to be adding TVF support to EF in general.

Does this work with EntityFramework6.Npgsql ?

neisbut commented:
@rwasef1830, oh, sorry, I forgot to check this. I made a mistake in function declaration. Fixed now. Could you please try again?

Actually my main goal is to support exactly EntityFramework.Functions :) For now it works for scalar function, but not for TVF. EntityFramework.Functions just generates function import code, but Npgsql still need to translate it to SQL. This is what I've implemented in this PR.

rwasef1830 commented on test/EntityFramework6.Npgsql.Tests/Support/EntityFrameworkTestBase.cs:
```diff
@@ -213,6 +223,85 @@ private static DbCompiledModel CreateModel(NpgsqlConnection connection)
+                {
+                    ParameterTypeSemantics = ParameterTypeSemantics.AllowImplicitConversion,
+                    Schema = "dbo",
+                    IsComposable = true,
```
Shouldn't this be IsComposable = false ? Or you can use TVF inside LINQ ? (I'm not familiar with TVFs in general, so I'm just making sure).

neisbut replied:
@rwasef1830, that's right, this flag allows to use this inside of LINQ expressions like this:
`var query2 = from b in context.GetBlogsByName("blog1")
                            select new { b.Name, Something = 1 };`

rwasef1830 commented on test/EntityFramework6.Npgsql.Tests/Support/EntityFrameworkTestBase.cs:
```diff
@@ -213,6 +223,85 @@ private static DbCompiledModel CreateModel(NpgsqlConnection connection)
+                new EdmFunctionPayload
+                {
+                    IsFunctionImport = true,
+                    IsComposable = true,
```
Same question here

neisbut replied:
@rwasef1830, yep, needed here as well.

rwasef1830 commented on test/EntityFramework6.Npgsql.Tests/Support/EntityFrameworkTestBase.cs:
```diff
@@ -213,6 +223,85 @@ private static DbCompiledModel CreateModel(NpgsqlConnection connection)
+                null);
+            dbModel.ConceptualModel.Container.AddFunctionImport(getBlogsFuncModel);
+
+            dbModel.ConceptualToStoreMapping.AddFunctionImportMapping(new System.Data.Entity.Core.Mapping.FunctionImportMappingComposable(
```
Line length is a little too long, can you use a "using" and some formatting and make this a bit more readable ?

rwasef1830 commented on test/EntityFramework6.Npgsql.Tests/Support/EntityFrameworkTestBase.cs:
```diff
@@ -213,6 +223,85 @@ private static DbCompiledModel CreateModel(NpgsqlConnection connection)
+            dbModel.ConceptualToStoreMapping.AddFunctionImportMapping(new System.Data.Entity.Core.Mapping.FunctionImportMappingComposable(
+                    getBlogsFuncModel,
+                    getBlogsFunc,
+                    new System.Data.Entity.Core.Mapping.FunctionImportResultMapping(),
```
Same comment, use namespace "using"

rwasef1830 commented on test/EntityFramework6.Npgsql.Tests/Support/EntityFrameworkTestBase.cs:
```diff
@@ -213,6 +223,85 @@ private static DbCompiledModel CreateModel(NpgsqlConnection connection)
+                    IsComposable = true,
+                    Parameters = new[] 
+                    {
+                        FunctionParameter.Create("Name", PrimitiveType.GetEdmPrimitiveTypes().First(x=>x.ClrEquivalentType == typeof(string)), ParameterMode.In)
```
Nitpick: spacing around "=>" in ```x=>x.ClrEquivalentType```. Also, the line is too long and it is not readable.
Perhaps pull all edm types into local variables above your function declarations and move it into a separate function to keep all the code for setting up this function together.

rwasef1830 commented on src/EntityFramework6.Npgsql/SqlGenerators/VisitedExpression.cs:
```diff
@@ -526,7 +526,7 @@ internal override void WriteSql(StringBuilder sqlText)
             else
             {
-                var wrap = !(_from is LiteralExpression || _from is ScanExpression);
+                var wrap = !(_from is LiteralExpression || _from is ScanExpression || _from is FunctionExpression);
```
Is avoiding the wrapping by parens necessary ? I'm not familiar with the pgsql for calling TVF sprocs so just making sure.

neisbut replied:
@rwasef1830, yes, I think this is necessary because of Postgres syntax. In example, the following works:
`select * from "dbo"."GetBlogsByName"('some')` 
but this one does not: 
`select * from ("dbo"."GetBlogsByName"('some'))`

rwasef1830 commented:
@neisbut The test passes now, just a few questions, formatting nits left.

neisbut commented on test/EntityFramework6.Npgsql.Tests/Support/EntityFrameworkTestBase.cs:
```diff
@@ -213,6 +224,87 @@ private static DbCompiledModel CreateModel(NpgsqlConnection connection)
+                null);
+            dbModel.ConceptualModel.Container.AddFunctionImport(getBlogsFuncModel);
+
+            dbModel.ConceptualToStoreMapping.AddFunctionImportMapping(new FunctionImportMappingComposable(
```
@rwasef1830, done, refactored long names.

rwasef1830 commented on test/EntityFramework6.Npgsql.Tests/Support/EntityFrameworkTestBase.cs:
```diff
@@ -213,6 +224,87 @@ private static DbCompiledModel CreateModel(NpgsqlConnection connection)
             dbModel.StoreModel.AddItem(echoFunc);
 
+            var stringStoreType = dbModel.ProviderManifest.GetStoreTypes().First(x => x.ClrEquivalentType == typeof(string));
+            var modelBlogStoreType = dbModel.StoreModel.EntityTypes.First(x => x.Name == typeof(Blog).Name) as EntityType;
```
Nit: "as EntityType" is redundant and generates a useless warning, you can safely remove it.

neisbut replied:
@rwasef1830, sure, no problem, done.

rwasef1830 commented on test/EntityFramework6.Npgsql.Tests/Support/EntityFrameworkTestBase.cs:
```diff
@@ -213,6 +224,87 @@ private static DbCompiledModel CreateModel(NpgsqlConnection connection)
+            dbModel.StoreModel.AddItem(getBlogsFunc);
+
+            var stringPrimitiveType = PrimitiveType.GetEdmPrimitiveTypes().First(x => x.ClrEquivalentType == typeof(string));
+            var modelBlogConceptualType = dbModel.ConceptualModel.EntityTypes.First(x => x.Name == typeof(Blog).Name) as EntityType;
```
Nit: "as EntityType" is redundant and generates a useless warning. You can safely remove it.

neisbut replied:
@rwasef1830, sure, no problem, done.

rwasef1830 commented:
@roji Please review this so I can merge it.
@neisbut Only a couple of nits and with @roji's approval it will be merged.

roji commented:
I definitely don't have time to dive into this at the moment... Formatting/cosmetics-wise things look good to me after @rwasef1830's comments, am OK for merging after they're fixed.

rwasef1830 commented:
@neisbut Thanks for contributing this feature!

neisbut commented:
@rwasef1830, guys, thank you very much! :)
